### ISSUE TITLE
clarification: rebar is a pcie 3.0 specification? all pcie 3.0 gpus are theoretically capable of full hw rebar?

### ISSUE THREAD

**Issue by zoulztealer** [terminatorul/NvStrapsReBar#34]
**System**
 - Motherboard: gigabyte x570 ud rev. 1.0
 - BIOS Version: f37 12/26/2022
 - GPU: zotac amp rtx 2080 ti tu102-300a rev. a1

**Description**

ive kept a close eye on rebar since it appeared and always waited for a mod like this.

now my questions are:

is rebar a pcie 3.0 specification?

can every pcie 3.0 gpu do full rebar?

why does capframex report full rebar in direct3d and does it mean i dont need an uefi / (v)bios mod to get the benefits of full rebar?

![rebar](https://github.com/terminatorul/NvStrapsReBar/assets/42149920/ae7a7d95-92a7-4000-8341-d622f0872e9c)

i have a zotac amp rtx 2080 ti here with a ryzen 5700x on a gigabyte x570 ud and 32gb 3600 dual rank ram. but since i dont have a vrr monitor yet (dont know if i will ever get one and rather get a 5800x3d for that price) and i run everything at 60hz/60fps @ 1080p ultra details im asking myself if i would benefit from this at all? would it improve latency, make things noticeable smoother or will it only help to achieve higher peak fps?

since ive optimized the bios and windows to the max everything runs butter smooth at max details so im very unsure if if should take the risk when there could be virtually no improvement also due to full rebar in d3d.

**Comment by terminatorul** [comment]
I could measure about 3% performance benefit in FPS, and 10-15% in PCI bandwidth test.

Other users measured 10% FPS increase in low-FPS max-quality (high texture) settings in Cyberpunk 2077 (but FPS was too low at that quality for the game to be playable, so the 10% increase is still not a real benefit).

**Comment by terminatorul** [comment]
Looks like PCI-SIG specifications are members-only and need to be purchased, but I can still see ReBAR introduced in apr 2008 for PCI v2.x. I might be wrong about it, though.

But the ReBAR extended capability is optional, and a lot of PCI devices that are not GPUs have no use for it. Definitely can say not every PCIe 3.0 GPU can do full rebar. For example my 1050 Ti uses PCI 3.0 but does not have the extended capability for ReBAR.

**Comment by terminatorul** [comment]
If you do not have ReBAR enabled in hardware on the PCI bus, I do not see how Resizable Bar D3D can show ON, or what that can even mean.

So if you are interested in performance, you should still enable it in the firmware, no matter what the capframex report shows

**Comment by zoulztealer** [comment]
thanks for your answers. that was exactly what i was looking for.

yes it seems to improve everything the gpu does. it would also slightly increase performance even in my case and i believe it would be even noticeable.

i would really like to go for it but im afraid it could fail and lose my windows installation or soft brick my board.

do you believe that sometime in the future there could be an easier / without modding or flashing a new bios or more automated way to activate it?

maybe something similar to what is mentioned here:

https://github.com/xCuri0/ReBarUEFI/wiki/Common-issues-(and-fixes)#how-do-i-enable-resizable-bar-on-unsupported-amd-gpus- (PCI-E Resizable Bar Service Mode and / regedit tweak)

its really strange what capframex shows there. it did the same for my gtx 1080 which i had before upgrading to the 2080 ti.

ive enabled above 4g and rebar in bios but nvcpl, gpuz and report that its turned off.

so i believe what capframex shows is the capability of direct3d to access full vram by default with some gpus. we would probably have to ask the guys from capframex what it is thats showing there.

**Comment by terminatorul** [comment]
Wait ... your board has Q-Flash Plus function. It means you can recover the board even if somehow you brick your board firmware. You're all set.

**Comment by zoulztealer** [comment]
ive asked in the tpu forums about the issue with capframex but it seems like noone knows about it. i wanted to inestigate it definitely though. think i will contact the guys from capframex and ask them about it.

it also seems to not be wholly identical with full hw rebar. maybe d3d can only access full vram but it comes with other limitations. but im not sure what this is really about.

ill read the whole process through and it seems pretty safe all over. ive also flashed custom x79 roms and other things before.

yes that board and that whole pc is pretty delicious. i also have other pcs and gpus here in case of an emergency. ive been tempted and waiting for so long! ;)

**Comment by zoulztealer** [comment]
can i just run the exe once and check on the current settings without making any changes?

**Comment by zoulztealer** [comment]
very cool.

![3333rebar](https://github.com/terminatorul/NvStrapsReBar/assets/42149920/af755b3b-0fa5-45bf-a2fa-6e9ad1a0364a)

**Comment by zoulztealer** [comment]
> Yes, sure

so ive tried and contacted capframex *twice* but i didnt get any reply what the partial rebar is and what is displayed here in the ui. pretty strange.

**Comment by terminatorul** [comment]
@terminatorul
>Next, the one rule you need to stick to, if you just want to be absolutely sure, is to disable ReBAR (using NvStrapsReBar.exe) before making changes in UEFI Setup and before making hardware changes. But even if you forget, you can recover by clearing CMOS and by reverting the hardware changes, or replacing the GPU with a different model.
>
>Or if you want to wait, I am still hoping to add checks to the DXE driver in a future version, for extra safety, to automatically disable ReBAR if settings have changed in UEFI Setup, or if hardware has changed.

First part is now implemented and, v0.4 will automatically disable NvStrapsReBar if you make changes in UEFI Setup.

This is a safety measure because changes in UEFI settings bring changes to an internal address saved by NvStrapsReBar when ReBAR is enabled (the GPU BAR0 base address). Simply enabling NvStrapsReBar again will also update the saved address, and thus prevent using the old value.

Also helpfull to prevent users from accidentally enable CSM or disable "Above 4G Decoding" option in UEFI Setup, when NvStrapsReBar is already enabled and depends on those options.

**Comment by zoulztealer** [comment]
> @terminatorul

bump

**Comment by zoulztealer** [comment]
> @terminatorul

strangely the gpuz advanced tab changed and 

gpu hardware support went from
"no" -> "yes"

resizeable bar enabled in bios went from
"unsupported gpu" to "no"

graphics driver support went from 
"unsupported gpu" to "yes"

i dont know how this happend because i didnt change anything except install a new driver.

![gpuz3](https://github.com/terminatorul/NvStrapsReBar/assets/42149920/2226de06-0e56-4804-969e-66e458bc99ee)

**Comment by zoulztealer** [comment]
so currently i am investigating a possibility to sideload the dxe uefi rebar module driver with a grub 2 compatible application such as ventoy and it seems possible indeed:

https://superuser.com/questions/1782009/can-grub-access-an-nvme-drive-when-the-bios-lacks-support-for-it

> [..] in the case of UEFI, it is possible to sideload a DXE driver that is compatible [..] (with EDK2 shell or Clover whatsoever), then your "(transiently) patched" UEFI firmware can load grub [..]

https://www.globalspec.com/reference/54257/203279/chapter-12-differences-between-dxe-drivers-and-efi-drivers

> The Extensible Firmware Interface (EFI) can be implemented in many ways. One way is to implement via the Driver Execution Environment (DXE) Foundation portion of the Framework. As such, DXE represents a special type of driver that can be combined with EFI drivers in a given firmware volume.
> 
> There are two basic classes of DXE drivers. The first class is DXE drivers that execute very early in the DXE phase. The execution order of these DXE drivers depends on the evaluation of dependency expressions. These early DXE drivers typically contain basic services, processor initialization code, chipset initialization code, and platform initialization code. These early drivers also typically produce the architectural protocols that are required for the DXE Core to produces its full complement of EFI Boot Services and EFI Runtime Services. In order to support that fastest possible boot time, as much initialization should be deferred to the DXE drivers that follow EFI Driver Model described in the EFI 1.10 Specification. Most of the platform and chipset drivers belong to this category. These drivers need to be aware that not all of these services may be available when they execute and use dependency expressions to make sure the protocols and services that they need are available.

**Comment by terminatorul** [comment]
I can imagine it is possible (but challenging!) to tell the GPU device to change BAR size after POST, from GRUB.

But this would also change the base address of the BAR. And then the GPU driver in the firmware is no longer in sync with the hardware. The firmware driver tries using the old address, and the hardware expects the new address.

Here the GPU driver is part of the system firmware, and is called GOP (Graphics Output Protocol). Windows has its own driver, and admittedly Windows would work even if the GOP does not. But only after boot. There is no boot progress, and if there is a boot problem and Windows can not load, or if you use a boot menu like GRUB, the screen would be frozen or corrupted.

In short, to change BAR size after POST means a new GOP driver should be provided as well. That would be a project in itself.

Then there is the issue of allocating the new base address, so that the new BAR does not conflict with any addresses already allocated for other devices in the system.

About GPU-Z, the maintainer added support for possible ReBAR on RTX2000 some time after NvStrapsReBar project showed up on github. I even provided some screenshots to them, after installing the pre-release (engineering) version of GPU-Z at the time.

**Comment by zoulztealer** [comment]
wow. very interesting. i never imagined it to be so difficult considered its supposed to be a pcie 2.0 / 2.1 old thing.

could it be possible to preprogram the *.ffs file / driver with everything from the current system like similar to what is being done with preprogramming the bar size etc right now with the NvStrapsReBar.exe?

for example i launch the exe while in windows 10 and it gathers everything from the running system and puts that in the ffs file driver. maybe this way things could be kept in sync between firm and hardware better by setting everything beforehand? maybe changes to the preconfiguration of the driver could be sufficient enough to allow the driver for being loaded more easily from grub or similar.

for example would it be possible to run a pci enumeration from windows 10 cmd line and store the information needed for booting inside the driver beforehand?

what also seems possible is that changes to rebar can be evoked by a graphic api like d3d. it would be critical to finding out what is reported in capframex with full rebar being available in d3d and how and why is this caused by d3d - and also why there is 214mb in vulkan vs 256mb in hw default. maybe this could reveal an easy software side hack that can be flipped on per game / api.

**Comment by dkchshv** [comment]
@zoulztealer could you please share your results? I have some minor results on a LEGACY system via DSDT modding but don't know how to go further. I modded DSDT directly in BIOS chip because I have two identical 8MB DIP8 chips and CH341B flash programmer.

This screenshot was taken in a portable Windows booted from Ventoy.

<img width="1024" height="726" alt="Image" src="https://github.com/user-attachments/assets/2865bacd-018e-4de7-b4bf-b196d092965e" />

```
$ sudo dmesg | grep "root bus resource" 
[    0.335336] pci_bus 0000:00: root bus resource [io  0x0000-0x0cf7 window]
[    0.335339] pci_bus 0000:00: root bus resource [io  0x0d00-0xffff window]
[    0.335341] pci_bus 0000:00: root bus resource [mem 0x000a0000-0x000bffff window]
[    0.335343] pci_bus 0000:00: root bus resource [mem 0x000d0000-0x000dffff window]
[    0.335346] pci_bus 0000:00: root bus resource [mem 0xde000000-0xf7ffffff window]
[    0.335348] pci_bus 0000:00: root bus resource [mem 0xfc000000-0xfed8ffff window]
[    0.335350] pci_bus 0000:00: root bus resource [mem 0x800000000-0xfffffffff window]
[    0.335352] pci_bus 0000:00: root bus resource [bus 00-ff]
[    0.345771] pci_bus 0000:3f: root bus resource [io  0x0000-0xffff]
[    0.345772] pci_bus 0000:3f: root bus resource [mem 0x00000000-0xfffffffff]
```

```
$ sudo cat /proc/iomem | grep -i "PCI Bus"
000a0000-000bffff : PCI Bus 0000:00
000d0000-000dffff : PCI Bus 0000:00
de000000-f7ffffff : PCI Bus 0000:00
  de000000-de1fffff : PCI Bus 0000:09
  de200000-de3fffff : PCI Bus 0000:0a
  e0000000-f4efffff : PCI Bus 0000:01
  f4f00000-f4ffffff : PCI Bus 0000:0b
  f6000000-f7afffff : PCI Bus 0000:01
  f7b00000-f7bfffff : PCI Bus 0000:02
  f7c00000-f7cfffff : PCI Bus 0000:0b
  f7d00000-f7ffffff : PCI Bus 0000:03
    f7e00000-f7ffffff : PCI Bus 0000:04
      f7e00000-f7efffff : PCI Bus 0000:05
      f7f00000-f7ffffff : PCI Bus 0000:07
fc000000-fed8ffff : PCI Bus 0000:00
800000000-fffffffff : PCI Bus 0000:00
  800000000-8001fffff : PCI Bus 0000:09
  800200000-8003fffff : PCI Bus 0000:0a
```

```
$ lspci -vv | grep -E "800000000|800200000" -B 10

00:1c.4 PCI bridge: Intel Corporation 5 Series/3400 Series Chipset PCI Express Root Port 5 (rev 06) (prog-if 00 [Normal decode])
        Subsystem: ASUSTeK Computer Inc. Device 8383
        Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR+ FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0, Cache Line Size: 32 bytes
        Interrupts: pin B routed to IRQ 17
        Bus: primary=00, secondary=09, subordinate=09, sec-latency=0
        I/O behind bridge: 1000-1fff [size=4K] [16-bit]
        Memory behind bridge: de000000-de1fffff [size=2M] [32-bit]
        Prefetchable memory behind bridge: 800000000-8001fffff [size=2M] [32-bit]
--

00:1c.5 PCI bridge: Intel Corporation 5 Series/3400 Series Chipset PCI Express Root Port 6 (rev 06) (prog-if 00 [Normal decode])
        Subsystem: ASUSTeK Computer Inc. Device 8383
        Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR+ FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0, Cache Line Size: 32 bytes
        Interrupts: pin C routed to IRQ 16
        Bus: primary=00, secondary=0a, subordinate=0a, sec-latency=0
        I/O behind bridge: 2000-2fff [size=4K] [16-bit]
        Memory behind bridge: de200000-de3fffff [size=2M] [32-bit]
        Prefetchable memory behind bridge: 800200000-8003fffff [size=2M] [32-bit]
```

**Comment by dkchshv** [comment]
@terminatorul @xCuri0 could you please take a look? : https://github.com/terminatorul/NvStrapsReBar/issues/34#issuecomment-4353162118

**Comment by dkchshv** [comment]
This could be notable too. Depending in which x16-lengh slot GPU is installed the bridges `00:03.0` and `00:05.0` change the show of 64-bit prefetchable memory, and this `[disabled] 64-bit` is quite similar to 2nd gen systems as I checked on another machine with 2nd generation Intel CPU:

```
$ sudo lspci -vvv -s 00:03.0
00:03.0 PCI bridge: Intel Corporation Core Processor PCI Express Root Port 1 (rev 11) (prog-if 00 [Normal decode])
        Subsystem: ASUSTeK Computer Inc. Device 8383
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR+ FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0, Cache Line Size: 32 bytes
        Interrupts: pin B routed to IRQ 16
        Bus: primary=00, secondary=01, subordinate=01, sec-latency=0
        I/O behind bridge: f000-0fff [disabled] [16-bit]
        Memory behind bridge: f5f00000-f5ffffff [size=1M] [32-bit]
        Prefetchable memory behind bridge: 00000000fff00000-00000000000fffff [disabled] [64-bit]
        Secondary status: 66MHz- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- <SERR- <PERR-
        BridgeCtl: Parity- SERR+ NoISA- VGA- VGA16- MAbort- >Reset- FastB2B-
                PriDiscTmr- SecDiscTmr- DiscTmrStat- DiscTmrSERREn-
        Capabilities: [40] Subsystem: ASUSTeK Computer Inc. Device 8383
        Capabilities: [60] MSI: Enable- Count=1/2 Maskable+ 64bit-
                Address: 00000000  Data: 0000
                Masking: 00000000  Pending: 00000000
        Capabilities: [90] Express (v2) Root Port (Slot+), IntMsgNum 0
                DevCap: MaxPayload 256 bytes, PhantFunc 0
                        ExtTag+ RBE+ TEE-IO-
                DevCtl: CorrErr- NonFatalErr- FatalErr- UnsupReq-
                        RlxdOrd- ExtTag+ PhantFunc- AuxPwr- NoSnoop-
                        MaxPayload 128 bytes, MaxReadReq 128 bytes
                DevSta: CorrErr- NonFatalErr- FatalErr- UnsupReq- AuxPwr- TransPend-
                LnkCap: Port #0, Speed 5GT/s, Width x16, ASPM L0s L1, Exit Latency L0s <512ns, L1 <4us
                        ClockPM- Surprise+ LLActRep+ BwNot+ ASPMOptComp-
                LnkCtl: ASPM Disabled; RCB 64 bytes, LnkDisable- CommClk-
                        ExtSynch- ClockPM- AutWidDis- BWInt+ AutBWInt+ FltModeDis-
                LnkSta: Speed 5GT/s, Width x4
                        TrErr- Train- SlotClk+ DLActive+ BWMgmt- ABWMgmt-
                SltCap: AttnBtn- PwrCtrl- MRL- AttnInd- PwrInd- HotPlug- Surprise-
                        Slot #1, PowerLimit 25W; Interlock- NoCompl-
                SltCtl: Enable: AttnBtn- PwrFlt- MRL- PresDet- CmdCplt- HPIrq- LinkChg-
                        Control: AttnInd Off, PwrInd Off, Power- Interlock-
                SltSta: Status: AttnBtn- PowerFlt- MRL- CmdCplt- PresDet- Interlock-
                        Changed: MRL- PresDet- LinkState-
                RootCap: CRSVisible+
                RootCtl: ErrCorrectable- ErrNon-Fatal- ErrFatal- PMEIntEna- CRSVisible+
                RootSta: PME ReqID 0000, PMEStatus- PMEPending-
                DevCap2: Completion Timeout: Range BCD, TimeoutDis+ NROPrPrP- LTR-
                         10BitTagComp- 10BitTagReq- OBFF Not Supported, ExtFmt- EETLPPrefix-
                         EmergencyPowerReduction Not Supported, EmergencyPowerReductionInit-
                         FRS- LN System CLS Not Supported, TPHComp- ExtTPHComp- ARIFwd+
                         AtomicOpsCap: Routing- 32bit- 64bit- 128bitCAS-
                DevCtl2: Completion Timeout: 260ms to 900ms, TimeoutDis- ARIFwd-
                         AtomicOpsCtl: ReqEn- EgressBlck-
                         IDOReq- IDOCompl- LTR- EmergencyPowerReductionReq-
                         10BitTagReq- OBFF Disabled, EETLPPrefixBlk-
                LnkCtl2: Target Link Speed: 5GT/s, EnterCompliance- SpeedDis-
                         Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
                         Compliance Preset/De-emphasis: -6dB de-emphasis, 0dB preshoot
                LnkSta2: Current De-emphasis Level: -6dB, EqualizationComplete- EqualizationPhase1-
                         EqualizationPhase2- EqualizationPhase3- LinkEqualizationRequest-
                         Retimer- 2Retimers- CrosslinkRes: unsupported, FltMode-
        Capabilities: [e0] Power Management version 3
                Flags: PMEClk- DSI- D1- D2- AuxCurrent=0mA PME(D0+,D1-,D2-,D3hot+,D3cold+)
                Status: D0 NoSoftRst+ PME-Enable- DSel=0 DScale=0 PME-
        Capabilities: [100 v1] Advanced Error Reporting
                UESta:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP-
                        ECRC- UnsupReq- ACSViol- UncorrIntErr- BlockedTLP- AtomicOpBlocked- TLPBlockedErr-
                        PoisonTLPBlocked- DMWrReqBlocked- IDECheck- MisIDETLP- PCRC_CHECK- TLPXlatBlocked-
                UEMsk:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP-
                        ECRC- UnsupReq- ACSViol- UncorrIntErr- BlockedTLP- AtomicOpBlocked- TLPBlockedErr-
                        PoisonTLPBlocked- DMWrReqBlocked- IDECheck- MisIDETLP- PCRC_CHECK- TLPXlatBlocked-
                UESvrt: DLP+ SDES+ TLP- FCP+ CmpltTO- CmpltAbrt- UnxCmplt- RxOF+ MalfTLP+
                        ECRC- UnsupReq- ACSViol- UncorrIntErr- BlockedTLP- AtomicOpBlocked- TLPBlockedErr-
                        PoisonTLPBlocked- DMWrReqBlocked- IDECheck- MisIDETLP- PCRC_CHECK- TLPXlatBlocked-
                CESta:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- AdvNonFatalErr- CorrIntErr- HeaderOF-
                CEMsk:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- AdvNonFatalErr+ CorrIntErr- HeaderOF-
                AERCap: First Error Pointer: 00, ECRCGenCap- ECRCGenEn- ECRCChkCap- ECRCChkEn-
                        MultHdrRecCap- MultHdrRecEn- TLPPfxPres- HdrLogCap-
                HeaderLog: 00000000 00000000 00000000 00000000
                RootCmd: CERptEn- NFERptEn- FERptEn-
                RootSta: CERcvd- MultCERcvd- UERcvd- MultUERcvd-
                         FirstFatal- NonFatalMsg- FatalMsg- IntMsgNum 0
                ErrorSrc: ERR_COR: 0000 ERR_FATAL/NONFATAL: 0000
        Capabilities: [150 v1] Access Control Services
                ACSCap: SrcValid+ TransBlk+ ReqRedir+ CmpltRedir+ UpstreamFwd+ EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
        Capabilities: [160 v0] Vendor Specific Information: ID=0002 Rev=0 Len=00c <?>
        Kernel driver in use: pcieport
        Kernel modules: shpchp
```

**Comment by zoulztealer** [comment]
sorry, i cant give you any results because i gave up on rebar after i created this post. simply not worth all the danger, specially since rebar works anyway, specially in d3d12 and hags it becomes much less important.

but i sincerely hope that someday it becomes an easy thing to switch on and off and controlled for all gpus, because thats what it should be like according to the specs.

**Comment by dkchshv** [comment]
> sorry, i cant give you any results because i gave up on rebar after i created this post. simply not worth all the danger, specially since rebar works anyway, specially in d3d12 and hags it becomes much less important.
> 
> but i sincerely hope that someday it becomes an easy thing to switch on and off and controlled for all gpus, because thats what it should be like according to the specs.

@zoulztealer ok. Thanks for answer anyway. I like your writings here. Something daring and poetic about it. They should've been just enable it by default in 2006-2007 or something instead of early 2020s because [my picture](https://private-user-images.githubusercontent.com/98752769/586042909-2865bacd-018e-4de7-b4bf-b196d092965e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzc2NjkwODgsIm5iZiI6MTc3NzY2ODc4OCwicGF0aCI6Ii85ODc1Mjc2OS81ODYwNDI5MDktMjg2NWJhY2QtMDE4ZS00ZGU3LWI0YmYtYjE5NmQwOTI5NjVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNTAxVDIwNTMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRkMTZkMjEyOTU2OTEyZGUyMWE5ODQ3OTgwMGMxZGVjYmU4OTI4YjViNmMzZTI3YjM3N2RhNDcwYjg5OTA5ZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.Fuw9bbTWAsPt8objdymdhdZ4AIXCK7gMUEndiOgXhQ0) with a 16+ years old CPU from 2009 literally shows a contradictions between what silicon could do but doesn't performs because of software/firmware which controls it. Absurdly just a memory hole for large BARs decoded above 4GB and GPU which's vRAM could be placed there. And artificial limitations with bunch of NO.

**Comment by xCuri0** [comment]
alot of ppl have said they want a way to be able to enable rebar without modding bios, it is technically possible to resize BARs, reallocate bridges/BARs, etc after POST (linux does it) but no one has written anything to do it.

**Comment by zoulztealer** [comment]
yes, thats how it should have been from the beginning according to pcie specs. thats why i said all these products are in violation of the pcie spec and basically a fraud. in fact bar is always active since d3d9 or something probably but user is effectively prohibited from changing bar size or so called re-bar. 

i think the reason they did this was because with wrong bar size the whole system becomes instable, so they set a fixed size and prevented user from control.

and yes, too, i think it should be doable pretty easy with some kind of modded driver or boot flag. the silicon we got is way more capable than is admitted by firmwares, true. dont even get me started on the 12 tera hertz transistor from 1942.

weve been there. weve done that.
